airflow/models/trigger.py

```diff
@@ -0,0 +1,161 @@
+    @classmethod
+    def from_object(cls, trigger: BaseTrigger):
+        """
+        Alternative constructor that creates a trigger row based directly
+        off of a Trigger object.
+        """
+        classpath, kwargs = trigger.serialize()
+        return cls(classpath=classpath, kwargs=kwargs)
+
+    @classmethod
+    @provide_session
+    def bulk_fetch(cls, ids: List[int], session=None) -> Dict[int, "Trigger"]:
+        """
+        Fetches all of the Triggers by ID and returns a dict mapping
+        ID -> Trigger instance
+        """
+        return {obj.id: obj for obj in session.query(cls).filter(cls.id.in_(ids)).all()}
+
+    @classmethod
+    @provide_session
+    def clean_unused(cls, session=None):
+        """
+        Deletes all triggers that have no tasks/DAGs dependent on them
+        (triggers have a one-to-many relationship to both)
+        """
+        # Update all task instances with trigger IDs that are not DEFERRED to remove them
+        session.query(TaskInstance).filter(
+            TaskInstance.state != State.DEFERRED, TaskInstance.trigger_id.isnot(None)
+        ).update({TaskInstance.trigger_id: None})
+        # Get all triggers that have no task instances depending on them...
+        ids = [
+            x[0]
+            for x in (
+                session.query(cls.id)
+                .join(TaskInstance, cls.id == TaskInstance.trigger_id, isouter=True)
+                .group_by(cls.id)
+                .having(func.count(TaskInstance.trigger_id) == 0)
+            )
+        ]
+        # ...and delete them (we can't do this in one query due to MySQL)
+        session.query(Trigger).filter(Trigger.id.in_(ids)).delete(synchronize_session=False)
+
+    @classmethod
+    @provide_session
+    def submit_event(cls, trigger_id, event, session=None):
+        """
+        Takes an event from an instance of itself, and triggers all dependent
+        tasks to resume.
+        """
+        for task_instance in session.query(TaskInstance).filter(
+            TaskInstance.trigger_id == trigger_id, TaskInstance.state == State.DEFERRED
+        ):
+            # Add the event's payload into the kwargs for the task
+            next_kwargs = task_instance.next_kwargs or {}
+            next_kwargs["event"] = event.payload
+            task_instance.next_kwargs = next_kwargs
+            # Remove ourselves as its trigger
+            task_instance.trigger_id = None
+            # Finally, mark it as scheduled so it gets re-queued
+            task_instance.state = State.SCHEDULED
+
+    @classmethod
+    @provide_session
+    def submit_failure(cls, trigger_id, session=None):
+        """
+        Called when a trigger has failed unexpectedly, and we need to mark
+        everything that depended on it as failed.
+        """
+        session.query(TaskInstance).filter(
+            TaskInstance.trigger_id == trigger_id, TaskInstance.state == State.DEFERRED
+        ).update({TaskInstance.state: State.FAILED})
```

Comment:
This won't call on_failure_callbacks, nor will it cascade to downstream tasks (which should end up in state upstream_failed) so this will need to be a bit more complex here.